### ISSUE TITLE
[Enhancement] Comic Book Reader UI Enhancements

### ISSUE THREAD

**Issue by ytilis** [janeczku/calibre-web#393]
While the kthoom comic book reader implemented works great, there are some issues from a usability standpoint I hope to resolve:

- No indication of what the keyboard shortcuts are (ie, fit to width, fit to height, rotate, etc)
- No way of panning or scrolling the page to see the rest of the page when fit to width on long pages
- No way to reliably jump to a specific page number

I propose implementing a number of UI elements which are already being used by ePubJS Reader to solve this. My suggestions are:

- Allow the canvas element to be scrolled on overflow, so that users can see the rest of the page when set to fit to width, etc
- Build a toolbar similar to the one used by the ebook reader with controls for existing functionality, and a modal to show keyboard shortcuts for these functions 
- Reuse the Table of Contents view from the ebook reader to show clickable page numbers and image thumbnails so that users can browse through them to specific parts of the book

I'm happy to make a pull request to implement this functionality, and would like to know if anyone has any feedback on these suggestions, or suggestions of their own while I get started.

**Comment by OzzieIsaacs** [comment]
Sounds good for me. Just for your information, i‘m Working on making the decompression of the comics on the server side, to give it a quicker start and to increase the overall performance, but I think both projects should not interfere with each other

**Comment by ytilis** [comment]
Glad to hear it. And yeah, this would purely be a frontend change, so the conflict should be pretty minimal.

Also, sorry if this is off-topic, but would that change you're working on be optional? I only ask because having it on the client now reduces server load and bandwidth use, and I'm just wondering what the impact would be of having the server decompress a file every time a user opens a comic or refreshes the page, especially on a smaller server or NAS.

**Comment by OzzieIsaacs** [comment]
I'm not so sure about the performance issue. I made some tests (lost my notes on the exact numbers). Before it was downloading a 60Mb comic (rar) and than unpacking it in about 100seconds (slow because of js/interpreted decomression instead of "real machine code" decompression). Now it is loading each page after each page, thus showing the first page within a few second and finishing after 60 seconds or so. Bandwith/transfer size: Because the jpeg picture format does a pretty good job in shrinking the page size the uncompressed size does not differ very much from the uncompressed size. 
Currently I do it this way: The page call requests the download of the first page, after this is archieved, the next page is requested. For limiting the bandwith and the server load, we could change it to just download the page which is currently shown (maybe one page ahead?). I created a new branch (comic_server) to let you feel the difference (error handling isn't completed, and I think I fiddled (?) on the progress bar). Please give Feedback

**Comment by ytilis** [comment]
Just pushed up the changes I've been working on for this ticket here:
https://github.com/ytilis/calibre-web/tree/cb-reader-ui

I *may* have gone a bit overboard, but the list of improvements include:
- Page style is now more consistent with EPUB reader 
- Table of Contents in left sidebar with image thumbnails to quickly jump to a specific page 
- Images are now scrollable when the fit size is changed 
- Fullscreen mode using the same logic and screenfull plugin from the EPUB reader (literally copied it from reader.min.js)
- Settings modal listing available keyboard shortcuts, as well as form controls to toggle them directly 
- A toggleable dark theme for comics

My last pet-peeve is the progress bar's garish colors and unnecessary use of SVG, since a div and some CSS would accomplish the same thing with way less code, but I figured I'd leave that alone since you mentioned you may have fiddled with it. 

----

RE server-side decompression, my main concern is the decompression to fetch the images in the first place to be honest, but looking at the code in your branch (and forgive me if I'm wrong, python isn't my strong suit), it looks like you're reading out individual files as needed, so my concern shouldn't be as much of a problem in that case. 

That said, one place it looks like we're definitely going to conflict is in the population of the pagination list I generate, since I currently insert the thumbnail into the TOC list when the image file is pushed into the array, but it looks like you've completely removed that section since there's no longer an extraction step.

**Comment by OzzieIsaacs** [comment]
I just merged both branches and commited it to the comic_server branch, feel free to check. The merge was easy, everthing fits together. 
Regarding the progressbar, I just checked it, I did nothing. So please feel free to make it a "nice one".

**Comment by ytilis** [comment]
Hmm, just pulled your branch and tried to open the comic reader. Got an internal server error:

```
[2017-11-19 12:30:19,235] {[...]/vendor/flask/app.py:1560} ERROR - Exception on /read/1/cbr [GET]
Traceback (most recent call last):
  File "[...]/vendor/flask/app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "[...]/vendor/flask/app.py", line 1615, in full_dispatch_request
    return self.finalize_request(rv)
  File "[...]/vendor/flask/app.py", line 1630, in finalize_request
    response = self.make_response(rv)
  File "[...]/vendor/flask/app.py", line 1725, in make_response
    raise ValueError('View function did not return a response')
ValueError: View function did not return a response

```
(I obfuscated my path prefix with `[...]` as an fyi)

**Comment by OzzieIsaacs** [comment]
Strange can't reproduce it. I did some minor changes, regarding return codes in error cases, you can try it, but I don't think it resolves the issue. 
Are there no other entries in the logfile? Is the file you are trying to open existing? Are you able to debug the python code? If yes, which line in "read_book(book_id, book_format)" (file web.py, lines 1917 to 1966) in leading to the error, means in which of those lines does the code return?

**Comment by ytilis** [comment]
I thought it might be because I didn't have ImageMagick installed, but that didn't seem to make a difference. Below is the whole calibre-web.log from starting the service to opening the file. If I'm on the master branch everything seems to work fine and I can read the book:

```
[2017-11-19 16:15:50,461] {[...]/cps/web.py:235} INFO - Starting Calibre Web...
[2017-11-19 16:15:50,594] {[...]/cps/book_formats.py:32} WARNING - cannot import epub, extracting epub metadata will not work: No module named lxml
[2017-11-19 16:15:50,603] {[...]/cps/book_formats.py:39} WARNING - cannot import fb2, extracting fb2 metadata will not work: No module named lxml
[2017-11-19 16:15:50,806] {cps.py:31} INFO - Falling back to Tornado
[2017-11-19 16:16:03,738] {[...]/vendor/flask/app.py:1560} ERROR - Exception on /read/1/cbr [GET]
Traceback (most recent call last):
  File "[...]/vendor/flask/app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "[...]/vendor/flask/app.py", line 1615, in full_dispatch_request
    return self.finalize_request(rv)
  File "[...]/vendor/flask/app.py", line 1630, in finalize_request
    response = self.make_response(rv)
  File "[...]/vendor/flask/app.py", line 1725, in make_response
    raise ValueError('View function did not return a response')
ValueError: View function did not return a response
```

Unfortunately I don't have the time to debug it right now but I'll try to figure out the specific line either later tonight or tomorrow if I can.

**Comment by andy29485** [comment]
Minor issue with the UI enhancements:
The arrow keys (left/right) change the page as expected, but when in `fit to width` mode
- up/down arrows do not scroll the page
- space does not scroll the page(down)
  - shift space does not scroll up (as it does with most web pages)

It would also be nice if:
- Hitting space, when at the bottom of the page, would change the page (same with down arrow maybe)
- Hitting shift+space, when at the top of the page, would change the page (back one)

**Comment by ytilis** [comment]
Figured out the issue I was having with the 500 error: 
`rar_support` was false for the condition on `web.py:1957`, which meant that `cbr` wasn't in the `extensionList`, even though that's the file type I was trying to read. So it looped through cbt and cbz, neither matches, and it just fails as a result. Of course removing the condition results in the page displaying, along with the message "Unable to decompress image #1" in the canvas. In the meantime, I can just use a cbz files to continue testing stuff. 

As for the issues you mentioned @andy29485, it looks like that's happening just because the body isn't the scroll window (since otherwise it'd mess with the toolbar and other stuff). But if you click on the image, it becomes the scroll area, allowing you to use space and arrow keys as expected. I'll see if I can have that happen automatically though, thanks for the observation.

**Comment by OzzieIsaacs** [comment]
To make cbr (rar files) work, you need to install rarfile ( pip...) and furthermore in the configuration page you need to point to unrar binary (hopefully installed, otherwise it needs to be installed) this is the only point I‘m not happy about. The error should now be prevented.

**Comment by OzzieIsaacs** [comment]
I had some time to work on it. Here is the outcome:
![sc1](https://user-images.githubusercontent.com/6438895/33475998-a0e2d8ea-d680-11e7-9a09-99a01046d713.jpg)
![sc2](https://user-images.githubusercontent.com/6438895/33476000-a10206ac-d680-11e7-988a-3319950b8dbe.jpg)
It's now also working on IE11 (which it does not before)

**Comment by ytilis** [comment]
Sorry for the delay, work got busy and I didn't have time to reconcile your changes with mine for a while.

Anyway, I had some time to work on it as well and added: 
- A slimmer and more subtle progress bar to the top which only shows the page number on hover 
- Current page indicator to the thumbnail list, as well as a hover state 
- Thumbnail list now scrolls to the current page on open or page change 
- Added @andy29485's request for controlling page with space/shift+space 
- Added the book title to the top of the page 
- And finally, I saw the IE 11 issue you mentioned and got it working. 

The problem with IE 11 was that it was seeing the server response as a string. Fun fact: `loadFromArrayBuffer` actually took this into account, but `ImageLoadCallback` didn't. So I just moved the json parse call further up the chain, and it seems to work now. Though it does hang when loading all the images, so the loading bar doesn't always advance as a result. May need to throttle that for IE possibly.

Made a new PR for now: #406
